agent/eventhandler/task_handler.go

```diff
@@ -15,63 +15,80 @@ package eventhandler
 
 import (
 	"container/list"
+	"sync"
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the TaskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	*list.List            // list of *sendableEvents
+	sending    bool       // whether the list is already being handled
+	mu         sync.Mutex // Locks both the list and sending bool
+}
+
+type TaskHandler struct {
+	submitSemaphore utils.Semaphore       // Semaphore on the number of tasks that may be handled at once
+	taskMap         map[string]*eventList // arn:*eventList map so events may be serialized per task
+	taskMapMu       sync.RWMutex          // Lock for the taskMap
```

Comment:
If you modify the previous line as per my suggestion, `tasksToEventsLock` is a better name for this. In any case, let's change it from `taskMapMu` to something else, preferably with the phrase `lock` somewhere in the mix.